File: transpiler-ts/src/parser/ast.ts
export type Node =
  | Program
  | FunctionDecl
  | AsBinding
  | CallStmt
  | StringLiteral
  | IntLiteral
  | Identifier;

export interface Program {
  kind: 'Program';
  decls: Node[];
}

export interface FunctionDecl {
  kind: 'FunctionDecl';
  name: string;
  returnType: string;
  params: Param[];
  body: Node[];
}

export interface Param {
  type: string;
  name: string;
}

export interface AsBinding {
  kind: 'AsBinding';
  name: string;
  value: Node;
}

export interface CallStmt {
  kind: 'CallStmt';
  func: string;
  args: Node[];
}

export interface StringLiteral {
  kind: 'StringLiteral';
  value: string;
}

export interface IntLiteral {
  kind: 'IntLiteral';
  value: string;
}

export interface Identifier {
  kind: 'Identifier';
  name: string;
}

File: transpiler/parser/ast.go
package parser

// Node is the base interface for all AST nodes.
type Node interface {
	nodeMarker()
}

type Program struct {
	Decls []Node
}

func (p *Program) nodeMarker() {}

type FunctionDecl struct {
	Name       string
	ReturnType string // empty = void
	Params     []Param
	Body       []Node
}

func (f *FunctionDecl) nodeMarker() {}

type Param struct {
	Type string
	Name string
}

// AsBinding represents `name as expr` — type inferred from literal shape.
type AsBinding struct {
	Name  string
	Value Node
}

func (a *AsBinding) nodeMarker() {}

// CallStmt represents a function call used as a statement.
type CallStmt struct {
	Func string
	Args []Node
}

func (c *CallStmt) nodeMarker() {}

type StringLiteral struct {
	Value string
}

func (s *StringLiteral) nodeMarker() {}

type IntLiteral struct {
	Value string
}

func (i *IntLiteral) nodeMarker() {}

type Identifier struct {
	Name string
}

func (id *Identifier) nodeMarker() {}
